### PR TITLE
SAK-48321 Kernel/Portal - Exiting Role View Redirects Instructor Out of Site and Role Swap Defaults to Overview Page

### PR DESCRIPTION
**Pull request by kunaljaykam**
https://sakaiproject.atlassian.net/browse/SAK-48321

### CODE REVIEW

File: kernel/kernel-impl/src/main/java/org/sakaiproject/authz/impl/SakaiSecurity.java [hunk 1]
```diff
@@ -1041,6 +1043,16 @@ public void update(Observable o, Object obj) {
 				resetSecurityCache(site.getReference());
 			}
 		}
+
+		if (EVENT_ROLESWAP_EXIT.equals(event.getEvent()))
+		{
+		  try {
+		    restoreUser();
+		  } catch (SakaiException e) {
+		    log.warn("Security invalidation error when handling an event (" + event.getEvent() + "), for user " + event.getResource());
```
Comment (adrianfish) [inline kernel/kernel-impl/src/main/java/org/sakaiproject/authz/impl/SakaiSecurity.java:1052]:
```suggestion
		    log.warn("Security invalidation error when handling an event ({}), for user {}", event.getEvent(), event.getResource());
```

File: kernel/kernel-impl/src/main/java/org/sakaiproject/authz/impl/SakaiSecurity.java [hunk 2]
```diff
@@ -1122,6 +1143,26 @@ protected void impersonateUser(String userId) throws SakaiException {
...
+
+	protected void restoreUser() throws SakaiException {
+		if (originalUserId == null || originalUserEid == null) {
+			throw new SakaiException("Original user session information is missing");
+		}
+
+		Session sakaiSession = sessionManager().getCurrentSession();
+
+		// Restore the original user session
+		Vector saveAttributes = new Vector();
```
Comment (adrianfish) [inline kernel/kernel-impl/src/main/java/org/sakaiproject/authz/impl/SakaiSecurity.java:1157]: Unless you need Vector's built in synchronisation you should probably use a Collection here
Collection<String> saveAttributes = new ArrayList<>();

Comment (adrianfish) [reply kernel/kernel-impl/src/main/java/org/sakaiproject/authz/impl/SakaiSecurity.java:1157]: Ah, right. It has to be a vector ..

**Review by adrianfish — APPROVED** [review]
This looks good to me. A complex piece of work, well done. I'll merge it and we can get it all tested.

**Comment by kunaljaykam** [comment]
> This looks good to me. A complex piece of work, well done. I'll merge it and we can get it all tested.

Thanks, Adrian!

**Comment by mpellicer** [comment]
This is great, thanks @kunaljaykam our instructors complained about the same and never realized we have a jira for it.

**Comment by kunaljaykam** [comment]
Thanks @mpellicer , but I might have to do another PR. I just noticed, If it is the user's first time swapping the role, it will redirect to /portal and it works fine after that. I will investigate why its happening. 
`31-Jan-2024 06:41:03.207 INFO [http-nio-8080-exec-8] o.s.a.i.SakaiSecurity.impersonateUser Enter into Role-Swap mode, user 3fd9deeb-9a19-4b2e-ab86-1b10df772411 is impersonated
31-Jan-2024 06:41:03.301 INFO [http-nio-8080-exec-2] o.s.p.l.ProfilePrivacyLogicImpl.getPrivacyRecordForUser Created default privacy record for user: 3fd9deeb-9a19-4b2e-ab86-1b10df772411
31-Jan-2024 06:41:03.304 INFO [http-nio-8080-exec-2] o.s.p.l.ProfilePreferencesLogicImpl.getPreferencesRecordForUser Created default preferences record for user: 3fd9deeb-9a19-4b2e-ab86-1b10df772411`

File: kernel/kernel-impl/src/main/java/org/sakaiproject/authz/impl/SakaiSecurity.java [hunk 3]
```diff
@@ -1094,8 +1106,17 @@ protected User newMockupStudentInSite(Site site, String eid, String role) throws
 		return newUser;
 	}
 	
+	// Instance variables to store original user's session information
+	private String originalUserId;
+	private String originalUserEid;
```
Comment (ern) [inline kernel/kernel-impl/src/main/java/org/sakaiproject/authz/impl/SakaiSecurity.java:1111]: These class member variables are visible globally and should not store data that is session scoped.

Comment (kunaljaykam) [reply kernel/kernel-impl/src/main/java/org/sakaiproject/authz/impl/SakaiSecurity.java:1111]: It hadn't occurred to me before, but you're right. I'll make sure to address this in my upcoming commit for the role swap feature. Thanks for pointing it out, Earle.

**Pull request merged**